File: javascript/howSum.js
function howSum(targetSum, numbers, memo = {}) {
    if (targetSum in memo) return memo[targetSum]
    if (targetSum === 0) return []
    if (targetSum < 0) return null

    for (let num of numbers) {
        const remainderResult = howSum(targetSum - num, numbers)
        memo[targetSum] = remainderResult
        if (remainderResult) return [...remainderResult, num]
    }

    memo[targetSum] = null
    return null
}

console.log(howSum(300, [4, 7]))

File: kotlin/howSum.kt
fun howSum(targetSum: Int, numbers: List<Int>): List<Int>? {
    if (targetSum == 0) return emptyList<Int>()
    if (targetSum < 0) return null

    numbers.forEach() {
        val remainderResult: List<Int>? = howSum(targetSum - it, numbers)
        if (remainderResult != null) return remainderResult + it
    }

    return null;
}

fun main(vararg args: String): Unit {
    println(howSum(args[0].toInt(), mutableListOf<Int>(4, 3, 5)))
}